utils/estimated_histogram.hh

```diff
@@ -47,13 +47,302 @@
+        }
+        auto c = count();
+
+        if (!c) {
+            return 0; // no data
+        }
+
+        auto pcount = uint64_t(std::floor(c * quantile));
+        uint64_t elements = 0;
+        for (size_t i = 0; i < NumBuckets - 2; i++) {
+            if (_buckets[i]) {
+                elements += _buckets[i];
+                if (elements >= pcount) {
+                    return get_bucket_lower_limit(i);
+                }
+            }
+        }
+        return Max; // overflowed value is in the requested quantile
+    }
+
+    /*!
+     * \brief returns the mean histogram value (average of bucket offsets, weighted by count)
+     */
+    uint64_t mean() const {
+        double elements = 0;
+        uint64_t sum = 0;
+        for (size_t i = 0; i < NumBuckets - 1; i++) {
+            elements += _buckets[i];
+            sum += _buckets[i] * get_bucket_lower_limit(i);
+        }
+        return (sum + elements - 1) / elements;
+    }
+
+    /*!
+     * \brief returns the number of buckets;
+     */
+    size_t size() const {
+        return NumBuckets;
+    }
+
+    /*!
+     * \brief returns the total number of values inserted
+     */
+    uint64_t count() const {
+        uint64_t sum = 0L;
+        for (size_t i = 0; i < NumBuckets; i++) {
+            sum += _buckets[i];
+        }
+        return sum;
+    }
+
+    /*!
+     * \brief multiple all the buckets content in the histogram by a constant
+     */
+    approx_exponential_histogram& operator*=(double v) {
+        for (size_t i = 0; i < NumBuckets; i++) {
+            _buckets[i] *= v;
+        }
+        return *this;
+    }
+};
+
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+inline approx_exponential_histogram<Min, Max, NumBuckets> base_estimated_histogram_merge(approx_exponential_histogram<Min, Max, NumBuckets> a, const approx_exponential_histogram<Min, Max, NumBuckets>& b) {
+    return a.merge(b);
+}
+
+/*!
+ * \brief estimated histogram for duration values
+ * time_estimated_histogram is used for common task timing.
+ * It covers the range of 0.5ms to 33s with a 2 bits granularity.
+ *
+ * 512us, 640us, 768us, 896us, 1024us, 1280us, 1536us, 1792us, 2048us, 2560us...
+ */
+class time_estimated_histogram : public approx_exponential_histogram<512, 33554432, 65> {
+public:
+    using clock = std::chrono::steady_clock;
+    using duration = clock::duration;
+    using approx_exponential_histogram<512, 33554432, 65>::add;
+    time_estimated_histogram& merge(const time_estimated_histogram& b) {
+        approx_exponential_histogram<512, 33554432, 65>::merge(b);
+        return *this;
+    }
+
+    void add_micro(uint64_t n) {
+        add(n);
+    }
+
+    void add(const duration& latency) {
+        add_micro(std::chrono::duration_cast<std::chrono::microseconds>(latency).count());
+    }
+};
+
+inline time_estimated_histogram time_estimated_histogram_merge(time_estimated_histogram a, const time_estimated_histogram& b) {
```

Comment:
@amnonh  I think it should return a `time_estimated_histogram&` rather than a copy of the object, since it's
being called like:
```
                s->result = s->reduce(std::move(s->result), std::move(f.get0()));
```
by `seastar::map_reduce`

So return an object rather than a reference causes 2 extra copies.